src/Algebra/Graph/Algorithm/Internal.hs

```diff
@@ -33,19 +34,22 @@ pointsList = mapM $ \(x, px) -> do
   repr <- UF.descriptor px
   return (repr, [x])
 
--- | O(s).
-mkVertexPoints :: Eq a => Graph a -> ST s ([(a, Point s a)], (Graph (Point s a)))
-mkVertexPoints Empty = return ([], Empty)
-mkVertexPoints (Vertex v) = do
-  point <- UF.fresh v
-  return ([(v, point)], Vertex point)
-mkVertexPoints (Connect l r) = do
-  (lp, lg) <- mkVertexPoints l
-  (rp, rg) <- mkVertexPoints r
+-- | O(s + n log n).
+mkVertexPoints :: Ord a => Graph a -> Map a (Point s a) -> ST s (Map a (Point s a), Graph (Point s a))
+mkVertexPoints Empty _ = return (Map.empty, Empty)
+mkVertexPoints (Vertex v) ps =
+  case Map.lookup v ps of
+    Nothing -> do
+      p <- UF.fresh v
+      return (Map.insert v p ps, Vertex p)
+    _ -> return (ps, Empty)
+mkVertexPoints (Connect l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
   return (lp <> rp, Connect lg rg)
-mkVertexPoints (Overlay l r) = do
-  (lp, lg) <- mkVertexPoints l
-  (rp, rg) <- mkVertexPoints r
+mkVertexPoints (Overlay l r) ps = do
+  (lp, lg) <- mkVertexPoints l ps
+  (rp, rg) <- mkVertexPoints r ps
```

Comment:
I replaced `[(a, Point s a]` with `Map a (Point s a)`. I tried to do it inside the `State` monad but got compilation error. After it rewrote with explicit state passing. I will try to fix the compilation error soon.